text/0011-peer-chaincode-optimization.md

```diff
@@ -0,0 +1,265 @@
+
+[1]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2458605718
+[2]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2460109295
+[3]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2473626315
+
+
+### Proposed Solutions
+
+1. Maintaining Read-Write Order
+
+	•	Solution: Ensure that the batching mechanism processes operations in the same order as they are invoked in the chaincode. This approach can be achieved by batching only consecutive write operations and sending read operations interactively.
+	•	Implementation: The chaincode shim can collect PutState or DelState operations into a batch, but any GetState operation will immediately flush the current batch and process the read interactively. This guarantees consistency with the current behavior.
+
+2. Handling Dependencies and Validation
+
+	•	Solution: Retain validation logic (e.g., metadata reads and collection name checks) on the peer side to ensure correctness. Avoid shifting these responsibilities to the shim, as it increases complexity and risk.
+	•	Implementation: The peer should process batched operations in a way that preserves existing checks and validations, ensuring no functional changes to current read-write dependencies.
+
+3. Explicit Developer Control
+
+	•	Solution: Introduce explicit APIs for developers to enable and manage batching in their chaincode. This gives developers control over when and how batching is applied, allowing for easier debugging and incremental adoption.
+	•	Implementation: APIs like StartBatch and FinishBatch can provide clear demarcation points for batching operations, ensuring developers understand and control the batching process.
+
+4. Optimizing Interleaved Operations
+
+	•	Solution: Allow partial batching where consecutive writes are batched, and reads or other operations trigger immediate processing of pending batches. This minimizes communication overhead without significantly altering the behavior for interleaved read-write scenarios.
+	•	Implementation: The shim can dynamically manage batches and flush them when necessary, ensuring optimal performance while maintaining behavior consistency.
+
```

Comment:
I don't understand the difference between option 1 and 4. Not a big deal though, since we are moving forward with option 3.

Reply:
removed to avoid confusion